### PR TITLE
Add wildcard support for includeDir

### LINKED ISSUE #723: Can c_cpp_properties.json support wildcard paths?
The scripts I am using to build my projects are generating the dependencies in a folder. It could be very useful in this situation to be able to give a path like this `/my/path/*/include` to simplify `/my/path/dep1/include` `/my/path/dep2/include` `/my/path/dep3/include`

I would be pleased to have your opinion. :)

### PR DESCRIPTION
**Pull request by yne**
For example, using the following tree:

- rootUri/
  - sources/
    - main/
      - main.c
    - vendor/
      - khash.h
  - headers/
    - C001
      - a.h
    - C002
      - b.h

|The following directory | will give
|-------------------|---------------
| `sources/**` | plain old recursive path like before
| `headers/C*/`| dynamic path
| `missing/`   | reported as "not found"
| `missing/*`  | expanded as empty result (silent fail)

Fixes #723

### CODE REVIEW

**Comment by yne** [comment]
@microsoft-github-policy-service agree

**Comment by yne** [comment]
`npm run unitTests` succeed (Exit code: 0) :heavy_check_mark:

But it also output the following message:

```
Unable to resolve your shell environment: Unexpected exit code from spawned shell
```

**Comment by sean-mcmanus** [comment]
Thanks for submitting this. We've been busy reviewing other PRs this past week but we'll try to get around to review this next week.

**Review by sean-mcmanus — COMMENTED** [review]
Does this expand subdirectories on the right side, such that it could potentially generate a large number of directories? If so, can that behavior be suppressed in favor of our existing recursive includes support? We want to avoid that because it negatively effects performance of our parser. It's okay for "internal" wildcard directories, since those are not as likely to generate a large number of directories.

**Comment by yne** [comment]
The right `/**` expression is kept out of the *new* glob process.
So only the *old* one can see it and can process it.

However, if there are any *middle* `**` expression, they will be expanded recursively by this new expansion process :/

I understand your worries about the performance impact. luckily for us `fast-glob` provide a [globstar](https://github.com/mrmlnc/fast-glob#globstar) option to disable this `**` behaviour.

I've amended my commit to specify a `globstar: false` option.

**Comment by sean-mcmanus** [comment]
FYI, we've been unexpectedly busy and haven't had time to review this yet. Our current target would be 1.15.0, since 1.14.x is entering stabilization mode.

File: Extension/src/LanguageServer/configurations.ts
Comment (sean-mcmanus) [inline Extension/src/LanguageServer/configurations.ts]: I don't think forcedInclude should use the globPaths (forced includes are a specific file that gets inlined) -- can you remove that?

**Review by sean-mcmanus — DISMISSED** [review]

File: Extension/src/LanguageServer/configurations.ts
Comment (yne) [reply Extension/src/LanguageServer/configurations.ts]: Sure, Done.

**Review by sean-mcmanus — COMMENTED** [review]
We've been busy with the 1.14.x release. I may be able to review this next week or the week after.

**Review by sean-mcmanus — DISMISSED** [review]
I don't see this change actually working. First, the changes only appear to be in getErrorsForConfigUI which is only invoked when the configuration UI is loaded. But even when I do that, I get 0 results for 
```TypeScript
            "includePath": [
                "${workspaceFolder}/*/inc"
            ],
```
How are you testing your change?

Also, there might be an issue with an incorrect squiggle appearing in the json UI editor for c_cpp_properties.json.

**Comment by yne** [comment]
I indeed tested through the Configuration UI.
I'll try from other means (direct JSON configuration edit).

What kind of squiggle ?

**Comment by sean-mcmanus** [comment]
> I indeed tested through the Configuration UI. I'll try from other means (direct JSON configuration edit).
> 
> What kind of squiggle ?

I see "Cannot find "c:\<workspade folder>\\inc"
image

**Comment by yne** [comment]
that's interesting.
I didn't tested on windows so I didn't got the backslashed star case (`\*`).
maybe it's preventing the star expansion.

I'll try to find a windows machine to test.

thanks for the info. I'll keep you informed.

**Comment by gerlero** [comment]
@yne As someone subscribed to #723, any update on this?

**Comment by yne** [comment]
Hi @gerlero, the only windows machine I found was my company machine which has network restrictions (no github push/pull allowed)

I could circumvent they restrictions but that's risky.

**Comment by yne** [comment]
hi @gerlero , I'm strugling to find a windows platform.
I could use my company computer but that not ideal since it's packed with a strict firewall that prevent any pull/push to github

**Comment by sean-mcmanus** [comment]
If you have a theoretical fix we could test it for you. Or can you download a zip of repo-branch for use on the Windows machine and then just re-do the fix/push on your other machine?

**Comment by yne** [comment]
Current status of my corporate windows machine and it "Secure Web Gateway" (aka man in the middle HTTPS proxy) :

- `git clone` failed (tried ssh and https) => Connection reset (probably blocked by firewall)
- so downloading the zip + extract in my home
- installing node 18.16.0 LTS from msi (required me administrator privilege; which not much people have).
- yarn installing via `corepack prepare yarn@stable --activate` did not work even with setting additional certificates
- So I manually put yarn.js 3.5.1 in my /appdata/ + `corepack prepare yarn@3.5.1 --activate`
- From the Extension folder, running `yarn install` fail with the .log given bellow 
- pressing F5 while being in the Extension folder give me some kind of "yarn: no such command" that only seems to occur in PowerShell based terminal (cmd works)
 
```
# This file contains the result of Yarn building a package (cpptools@workspace:.)
# Script name: postinstall


> cpptools@1.14.3-main download-api
> vscode-dts dev

Downloading vscode.proposed. [32mterminalDataWriteEvent [0m.d.ts
To:   C:\Users\yne\vscode-cpptools-release\Extension\vscode.proposed.terminalDataWriteEvent.d.ts
From: [https://raw.githubusercontent.com/microsoft/vscode/main/src/vscode-dts/vscode.proposed.terminalDataWriteEvent.d.ts](https://www.google.com/url?q=https%3A%2F%2Fraw.githubusercontent.com%2Fmicrosoft%2Fvscode%2Fmain%2Fsrc%2Fvscode-dts%2Fvscode.proposed.terminalDataWriteEvent.d.ts&sa=D&sntz=1&usg=AOvVaw33lAJq5oGA3qzOxELyOTiz)
Read more about proposed API at: [https://code.visualstudio.com/api/advanced-topics/using-proposed-api](https://www.google.com/url?q=https%3A%2F%2Fcode.visualstudio.com%2Fapi%2Fadvanced-topics%2Fusing-proposed-api&sa=D&sntz=1&usg=AOvVaw3WjbXUCKa3m6RcBPUfe9Pp)
node:events:491
      throw er; // Unhandled 'error' event
      ^

Error: read ECONNRESET
    at TLSWrap.onStreamRead (node:internal/stream_base_commons:217:20)
Emitted 'error' event on ClientRequest instance at:
    at TLSSocket.socketErrorListener (node:_http_client:502:9)
    at TLSSocket.emit (node:events:513:28)
    at emitErrorNT (node:internal/streams/destroy:151:8)
    at emitErrorCloseNT (node:internal/streams/destroy:116:3)
    at process.processTicksAndRejections (node:internal/process/task_queues:82:21) {
  errno: -4077,
  code: 'ECONNRESET',
  syscall: 'read'
}

Node.js v18.16.0
```

Having battled with JavaScript ecosystem in the past, I'm not surprised having so much hard time doing a simple environment setup. However I was wondering if some of those issues are more windows or Node related ?
Am I close to have a working enviroment or shall I just give up and download a VM ?

**Comment by sean-mcmanus** [comment]
We're still using yarn 1.22.x and it may not work with yarn 3.5.

**Comment by yne** [comment]
It get stuck at the same step even with yarn 1.22

I tried running a W11 VM but it was unbearable on my 8Go laptop

So I finally got my hands on a dualboot PC and it seems to be working, and without a surprise the issue was that the windows separator `\` are considered as escape character. This makes fast-glob consider the path as dynamic (https://www.npmjs.com/package/fast-glob#what-is-a-static-or-dynamic-pattern), But it's fixed now:

image

I'll just need to get [the changes](https://github.com/microsoft/vscode-cpptools/files/11644457/src.zip) back into git.

**Comment by yne** [comment]
@sean-mcmanus I've rebased onto main and committed the fix in a separate commit for easier review.
It shall now work, but note that using glob with a slash-ending path may return an empty match

File: Extension/src/LanguageServer/configurations.ts
Comment (yne) [inline Extension/src/LanguageServer/configurations.ts]: In my case I had to use a **middle** recursive-glob `path/**/headers`
But note that **ending** recursive glob are processed like before (not by fastglob)

**Comment by yne** [comment]
lint fixed and rebased onto `main` :rocket:

**Comment by sean-mcmanus** [comment]
@yne Yeah, it's possible "main" requires changes in the binary that only available with 1.16.1 which will be released later today.

**Comment by sean-mcmanus** [comment]
@yne FYI, main is "locked down" for the 1.16.2 release (maybe for Thursday) until we merge to the release branch. I think this can get into 1.17.0 (pre-release).

**Review by fearthecowboy — APPROVED** [review]
I'm OK with this - I haven't tested the functionality.

This can get merged ahead of #11085

**Comment by browntarik** [comment]
After attempting to test the feature myself, the support for dynamic paths does not seem to work, as a result we will delay checking this in. The example configuration did not yield the correct include paths.
Configuration
{
    "configurations": [
        {
            "name": "Win32",
            "includePath": [
                "sources/nest*/**"
            ],
            "defines": [
                "_DEBUG",
                "UNICODE",
                "_UNICODE"
            ],
            "windowsSdkVersion": "10.0.22000.0",
            "cStandard": "c17",
            "cppStandard": "c++17",
            "intelliSenseMode": "windows-msvc-x64"
        }
    ],
    "version": 4
}

Log Output:
-------- Diagnostics - 6/26/2023, 3:23:19 PM
Version: 1.16.1-main
Current Configuration:
{
    "name": "Win32",
    "includePath": [],
    "defines": [
        "_DEBUG",
        "UNICODE",
        "_UNICODE"
    ],
    "windowsSdkVersion": "10.0.22000.0",
    "cStandard": "c17",
    "cppStandard": "c++17",
    "intelliSenseMode": "windows-msvc-x64",
    "compilerPathIsExplicit": true,
    "cStandardIsExplicit": true,
    "cppStandardIsExplicit": true,
    "intelliSenseModeIsExplicit": true,
    "mergeConfigurations": false,
    "compilerPath": "s",
    "browse": {
        "path": [
            "${workspaceFolder}"
        ],
        "limitSymbolsToIncludedHeaders": true
    }
}
cpptools version (native): 1.16.1.0
Translation Unit Mappings:
[ C:\Users\browntarik\Desktop\Test\sources\nested\nested.h ]:
    C:\Users\browntarik\Desktop\Test\sources\nested\nested.h
Translation Unit Configurations:
[ C:\Users\browntarik\Desktop\Test\sources\nested\nested.h ]:
    Process ID: 29076
    Memory Usage: 51 MB
    Compiler Path: C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC\Tools\MSVC\14.29.30133\bin\Hostx64\x64\cl.exe
    Includes:
        C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC\Tools\MSVC\14.29.30133\include
        C:\Program Files (x86)\Windows Kits\10\Include\10.0.22000.0\um
        C:\Program Files (x86)\Windows Kits\10\Include\10.0.22000.0\ucrt
        C:\Program Files (x86)\Windows Kits\10\Include\10.0.22000.0\shared
        C:\Program Files (x86)\Windows Kits\10\Include\10.0.22000.0\winrt
        C:\Program Files (x86)\Windows Kits\10\Include\10.0.22000.0\cppwinrt
    Defines:
        _DEBUG
        UNICODE
        _UNICODE
    Standard Version: ms_c++17
    IntelliSense Mode: windows-msvc-x64
    Other Flags:
        --header_only_fallback
Total Memory Usage: 51 MB

------- Workspace parsing diagnostics -------
Number of files discovered (not excluded): 5028

**Comment by sean-mcmanus** [comment]
@yne Can headers/C*/** be made to work? The expanded paths should just preserve the "**" at the end so we can process those. Or did you want to exclude that for some reason?

**Comment by yne** [comment]
the ending ** are kept out of glob process and re-added post-globbing

It worked on windows in my last fix but I cannot test it again now (see: my last braindead vscode issue)

Is there any formal verification procedure we could agreed upon ?

I'll let you tell me when I can start debugging again (after the problematic plugin migration?)

**Comment by browntarik** [comment]
The feature does work as long as "**" is excluded. I will make a note for that to be a future improvement and we can go ahead and check this in for now.

**Review by browntarik — APPROVED** [review]

**Comment by sean-mcmanus** [comment]
@browntarik @yne On Windows, `headers/C*/` doesn't work, but `headers/C*` works.

**Comment by sean-mcmanus** [comment]
@yne FYI, you don't need to fix this. I think @browntarik was going to (i.e. https://github.com/microsoft/vscode-cpptools/issues/11135 is assigned to him). The shipped 1.16.3 binaries should be usable for testing.

**Comment by haolly** [comment]
@sean-mcmanus I need this feature, how can I try it ASAP ?

**Comment by sean-mcmanus** [comment]
@haolly It should be available in our 1.17.0 release. We're not sure when that will be yet, because we're still working on fixing some issues.

**Pull request merged**
